engine/crates/runtime-local/src/rate_limiting.rs

```diff
@@ -0,0 +1,174 @@
+use std::collections::HashSet;
+use std::net::IpAddr;
+use std::num::NonZeroU32;
+use std::str::FromStr;
+
+use futures_util::future::{ready, BoxFuture};
+use futures_util::FutureExt;
+use governor::{DefaultKeyedRateLimiter, Quota};
+use tungstenite::http;
+
+use registry_v2::rate_limiting::{AnyOr, Header, Jwt, RateLimitRule, RateLimitRuleCondition};
+use runtime::rate_limiting::{Error, RateLimiter, RateLimiterContext};
+
+pub struct InMemoryRateLimiting {
+    rate_limiters: Vec<(RateLimitRuleCondition, DefaultKeyedRateLimiter<String>)>,
+}
+
+impl InMemoryRateLimiting {
+    pub fn new(rules: &[RateLimitRule]) -> Self {
+        Self {
+            rate_limiters: rules
+                .iter()
+                .map(|rule| {
+                    let quota = rule
+                        .limit
+                        .checked_div(rule.duration.as_secs() as u32)
+                        .expect("rate limiter with invalid per second quota");
+
+                    (
+                        rule.condition.clone(),
+                        governor::RateLimiter::keyed(Quota::per_second(
+                            NonZeroU32::new(quota).expect("rate limit duration cannot be 0"),
```

Comment:
Should the parser be enforcing this non-zero constraint as well?  Took a quick look and couldn't see it doing that, correct me if I just missed it